### PR TITLE
dns/dyndns: Implementend a basic PoC for Gandi LiveDNS

### PR DESCRIPTION
**Pull request by vizion8-dan**
As requested in #1743  - this is a working version of the [Gandi LiveDNS](https://api.gandi.net/docs/livedns/#v5-livedns-domains-fqdn-records-rrset_name-rrset_type) for OPNsense. 

Works for me. Personally, I can not provide an API key for testing at the moment, but have tested it succesfully with my own Gandi API key.

### CODE REVIEW

**Comment by vizion8-dan** [comment]
Merged with godaddy changes...

File: dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc [hunk 1]
```diff
@@ -669,6 +671,31 @@ class updatedns
...
+                * dnsPass should be the Gandi-API key
+                */
+                $server = "https://dns.api.gandi.net/api/v5/domains/". $this->_dnsHost . "/records/" . $this->_dnsUser . "/A";
+
+                log_error("DAN: ". $server);
+                $body = '{"rrset_ttl":"' . "300" . '", "rrset_values":["'. $this->_dnsIP . '"]}';
+                log_error("DAN: ". $body);
+
+                curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, true);
+                curl_setopt($ch, CURLOPT_CUSTOMREQUEST, 'POST');
```
[the remark this answers is not in the document]
Comment (vizion8-dan) [reply dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc:689]: Hi, thanks for the test and the feedback! I`ll have a look at the proposed change tomorrow but using PUT sounds good! Thanks 😃

[the remark this answers is not in the document]
Comment (vizion8-dan) [reply dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc:689]: Again, thanks for checking and finding this! It should have been PUT as in my https://github.com/opnsense/plugins/issues/1743#issuecomment-651038748 code snippet.

[the remark this answers is not in the document]
Comment (vizion8-dan) [reply dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc]: Corrected

**Comment by admlko** [comment]
LGTM! 👍

File: dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc [hunk 2]
```diff
@@ -42,6 +42,7 @@
  *    - regfish IPv6 (regfish.de)
  *    - dynv6 IPv6 (dynv6.com)
  *    - DigitalOcean (digitalocean.com)
+ *    - Gandi LiveDNS (gandi.net)
```
Comment (fichtner) [inline dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc:45]: There is also a "last tested" entry below. Can you fill in the date? :)

File: dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc [hunk 3]
```diff
@@ -669,6 +671,31 @@ class updatedns
...
+                /*
+                * https://github.com/vizion8-dan
+                * Tested on OPNsense 20.1.8_1-amd64 - IPv4 (A)
+                * dnsHost ("Hostname" field in OPNsense) should be the 2nd-level domain ("example.org") 
+                * dnsUser ("Username" field in OPNsense) should be the subdomain / A-record ("myrecord" in 2nd-level domain) 
+                * dnsPass should be the Gandi-API key
+                */
+                $server = "https://dns.api.gandi.net/api/v5/domains/". $this->_dnsHost . "/records/" . $this->_dnsUser . "/A";
+
+                log_error("DAN: ". $server);
```
Comment (fichtner) [inline dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc:684]: debug logs?

**Review by fichtner — APPROVED** [review]
ok minus comments. almost ready for merge.

**Comment by vizion8-dan** [comment]
Hi @fichtner could you please clarify whats missing? I think its about the log `log_error("DAN: ". $server);`, isn`t it?  - it can be removed as it was a helper for me to understand some mechanics. I am removing it.

**Comment by vizion8-dan** [comment]
Sorry, for that. I have found the comments.

**Comment by fichtner** [comment]
yep, please remove log_error() additions and add a last test date for the gandi livedns implementation in the header

File: dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc [hunk 4]
```diff
@@ -669,6 +671,29 @@ class updatedns
...
+                curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, true);
+                curl_setopt($ch, CURLOPT_CUSTOMREQUEST, 'PUT');
+                curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
+                curl_setopt($ch, CURLOPT_HTTPHEADER, array(
+                    "X-Api-Key: {$this->_dnsPass}",
+                    'Content-Type: application/json'
+                ));
+                curl_setopt($ch, CURLOPT_POSTFIELDS, $body);
+                curl_setopt($ch, CURLOPT_URL, $server);
+                log_error("DAN: ". $server);
```
Comment (fichtner) [inline dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc:695]: one more here :)

File: dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc (hunk 2 above)
Comment (vizion8-dan) [reply dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc:45]: Sure. Inserted it.

**Comment by vizion8-dan** [comment]
> yep, please remove log_error() additions and add a last test date for the gandi livedns implementation in the header

Done.

File: dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc (hunk 4 above)
Comment (vizion8-dan) [reply dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc:696]: Removed both.

Comment (fichtner) [reply dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc:696]: there is a third still in there

Comment (vizion8-dan) [reply dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc]: Sorry, missed this one. Removed.

**Comment by fichtner** [comment]
Merged, thank you ❤️

**Comment by admlko** [comment]
❤️

**Pull request merged**